src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,460 @@
+            NotificationSettingProvider.objects.filter(query & type_filter & provider_filter)
+        )
+
+    def get_all_setting_options(self) -> Iterable[NotificationSettingOption]:
+        return self._setting_options
+
+    def get_all_setting_providers(self) -> Iterable[NotificationSettingProvider]:
+        return self._setting_providers
+
+    def _get_query(self) -> Q:
+        """
+        Generates a query for all settings for a project, org, user, or team.
+
+        Args:
+            recipients: The recipients of the notification settings (user or team).
+            projects_ids: The projects to get notification settings for.
+            organization_id: The organization to get notification settings for.
+        """
+        if not self.recipients:
+            raise Exception("recipient, team_ids, or user_ids must be provided")
+
+        user_ids, team_ids = [], []
+        for recipient in self.recipients:
+            if recipient_is_user(recipient):
+                user_ids.append(recipient.id)
+            elif recipient_is_team(recipient):
+                team_ids.append(recipient.id)
+
+        if not user_ids and not team_ids:
+            raise Exception("recipients must be either user or team")
+
+        project_settings = (
+            Q(
+                (Q(user_id__in=user_ids) | Q(team_id__in=team_ids)),
+                scope_type=NotificationScopeEnum.PROJECT.value,
+                scope_identifier__in=self.project_ids,
+            )
+            if self.project_ids
+            else Q()
+        )
+
+        org_settings = (
+            Q(
+                (Q(user_id__in=user_ids) | Q(team_id__in=team_ids)),
+                scope_type=NotificationScopeEnum.ORGANIZATION.value,
+                scope_identifier=self.organization_id,
+            )
+            if self.organization_id
+            else Q()
+        )
+
+        team_or_user_settings = Q(
+            (Q(user_id__in=user_ids) | Q(team_id__in=team_ids)),
+            scope_type=NotificationScopeEnum.USER.value,
+            scope_identifier__in=user_ids,
+        )
+
+        return project_settings | org_settings | team_or_user_settings
+
+    def _get_scope_for_layering(
+        self, recipient: RpcActor | Team | RpcUser
+    ) -> Tuple[NotificationScopeEnum, Iterable[int | RpcActor | Team | RpcUser]]:
+        if self.project_ids:
+            return (NotificationScopeEnum.PROJECT, self.project_ids)
+        if self.organization_id is not None:
+            return (NotificationScopeEnum.ORGANIZATION, [self.organization_id])
+        if recipient_is_user(recipient):
+            return (NotificationScopeEnum.USER, [recipient])
+        if recipient_is_team(recipient):
+            return (NotificationScopeEnum.TEAM, [recipient])
+
+        raise ValueError(f"Recipient is of unsupported type: {recipient}")
+
+    def _filter_options(
+        self,
+        settings: Iterable[NotificationSettingOption],
```

Comment:
@snigdhas do we need this as an argument? Can't we just look at `self.get_all_setting_options`)?

Reply:
We filter the settings in multiple passes (i.e. split up the settings by scope). The param allows us to specify the already filtered settings without overwriting or recomputing `self.get_all_setting_options`.